### ISSUE TITLE
Use a new file to store the outlined function of target directive

### ISSUE THREAD

**Issue by ouankou** [passlab/rexompiler#64]
According to Dr. Yan's [study](https://github.com/passlab/rexruntime/tree/master/tests/libomptarget), to transform the `target` directive, we need a separate `.cu` file to store the CUDA kernel function. In the original file, no CUDA code is involved.

Currently, ROSE stores the outlined function in the same source file. There could be two solutions:

1. Manually create a new file, copy the relevant AST there.
2. ROSE may have built-in functions to store the outlined function in a new file.

These are only rough ideas. More details of ROSE source code should be explored.

**Comment by ouankou** [comment]
For outlining in OpenMP, ROSE enables `useParameterWrapper` and `useStructureWrapper` to include the variables into a data structure and pass it to the outlined function.

Another flag `useNewFile` is to store the outlined function in a new file. However, ROSE doesn't support using `useNewFile` and `useStructureWrapper` at the same time. Otherwise, it will complain that `Outliner::generateParameterStructureDeclaration() separated file case is not yet handled.`

The source code that throws the error message is: https://github.com/passlab/rexompiler/blob/master/src/midend/programTransformation/astOutlining/Transform.cc#L140

It means the **soultion 2** requires us to:
1. explore the details of error and implements any missing feature in ROSE, or
2. significantly change the workflow in omplowering to not using structure wrapper.

**Comment by ouankou** [comment]
## Solution 1:

So far, a new file with given name could be created with the outlined function. However, it seems only the function SgNode itself is copied but not its full subtree.

Given a simple input:

```c
#include <stdio.h>

int main (int argc, char** argv) {
      int i = 1;
        #pragma omp parallel if(i == 1)
      {
          #pragma omp single
            printf("Test\n");
      }
}
```

The regular transformed code is:

```c
#include <stdio.h>
#include "rex_kmp.h"
static void OUT__1__5599__(int *__global_tid,int *__bound_tid,void *__out_argv);

int main(int argc,char **argv)
{
  int __global_tid_value = __kmpc_global_thread_num(0);
  int *__global_tid = &__global_tid_value;
  int status = 0;
  int i = 1;
  if (i == 1)
    __kmpc_fork_call(0,1,OUT__1__5599__,0);
   else {
    OUT__1__5599__(&( *__global_tid),0,0);
  }
}

static void OUT__1__5599__(int *__global_tid,int *__bound_tid,void *__out_argv)
{
  if (__kmpc_single(0, *__global_tid)) {
    printf("Test\n");
    __kmpc_end_single(0, *__global_tid);
  }
  __kmpc_barrier(0, *__global_tid);
}
```

The current implementation of solution 1 produces a separated file:

```c
static void OUT__1__5599__(int *__global_tid,int *__bound_tid,void *__out_argv)
{
  if (__kmpc_single(0, *__global_tid)) {
    __kmpc_end_single(0, *__global_tid);
  }
  __kmpc_barrier(0, *__global_tid);
}
```

There are several problems:

1. The separated file also needs to include `rex_kmp.h`.
2. The modifier `static` in the separated file should be removed.
3. The modifier `static` in the main file should be changed to `extern`.
4. The separated file should have the complete outlined function definition.
5. The outlined function definition in the main file should be removed. Its SgNode may or may not be kept, but it shouldn't be unparsed for sure.
5. This simple case doesn't have a parameter structrue, but we expect the structure will encounter the similar issue with the outlined function.
6. Should we store all the outlined functions in one file and multiple files (one function per file)?

Problem 1 should be easy to fix. For problems 2 and 3, we may implement several outlining modules by ourselves because the forward declaration can't be tracked via the outliner API to change its storage modifier.

**Comment by yanyh15** [comment]
@ouankou Thank you for explore this in details. I could not find the code I created before, more than 10 years ago. I vaguely recall that it will need a deepcopy of AST in order to duplicate the AST. Did you try [sageInterface::deepCopyNode (const SgNode *subtree)](http://rosecompiler.org/ROSE_HTML_Reference/namespaceSageInterface.html#a2f77cb853d76a6bd75e034fb2da0c07d).  We can ask Liao for details. 

I am thinking of two options for files vs regions:  1) a new file for each outlined function, or 2) one file for all the outlined function. It seems both are ok for me. Considering that we will support offloading (CUDA GPU) and vectorization (simd instruction or intrinsics), we can have one file for each architecture (CPU, CUDA, simd) that includes the outlined function for all the regions of a file, or one file per outlined region per architecture. 

With either option, there are some details that need to be handled for the new file to be compilable, e.g. for the signatures of several things: 
1. header include, we can copy the original include to the new file. 
2. typedef, extern, #define can all be copied to the new file. 
3. global variable needs to be added to the new file as extern-ed variable
4. all the function used by the outlined function body needs to have correct function signature in the new file
5. others (?). There might be other corner cases. 

We can create a new header that include the above info and include this header in the file for the outline function, if we need to create several files for all the outline regions. This approach just simply copy all the signature definition of the original file. A better approach is just to extract the signature declaration of those symbols used in the outlined body, as well as the macro used.

**Comment by ouankou** [comment]
@yanyh15 Yes, I used the similar API `deepCopy`. I tried both and got the same full subtree. They are probably the same thing. Currently, the first 5 problems have been solved. I'm working on problem 7. For simple cases, it works fine. The function declarations are all copied to the new file and the two generated files can be compiled to get the correct executable.

However, I'm sure there will be issues for normal cases due to the points you listed. Additonally, since some ROSE APIs don't work in my code, there must be something wrong underneath. I'll wrap up the current implementation and push a commit. Then we can keep updating the design from there.

**Comment by yanyh15** [comment]
This is a general problem: how to split a file into multiple files. It is not an easy job if we need to do perfect because of the situation we need to consider: 
1.

**Comment by ouankou** [comment]
According to Dr. Liao, ROSE does keep all the definitions of global variables in the separate file having outlined functions. The compiler will complain that they are redeclared. We may need to manually scan those variables and mark them as extern declarations.

**Comment by ouankou** [comment]
There could be multiple issues when we create a new file to hold the outlined functions.

1. Global variables
The main file may have global variables. If they are static, they need to be copied. Otherwise, we have to set them to be extern without definition.
For example, `int a = 1;` should be `extern int a;`, but `static b = 2;` should be preserved as is.

2. Preprocessing `#define`
They should be copied exactly as is because they are not shared across source files.

3. Function declaration
Functions other than the outlined ones should only have the extern declaration in the new file if they are not static. Otherwise, they should be copied with the definition.
The outline functions should have the forward declaration in the new file in case they are calling each other.
In the main file, the outline functions only have the extern declaration there.
